### DISCUSSION TITLE
When we use `typescript-language-server`, do we need a Client?

### DISCUSSION THREAD

**Discussion by liaodalin19903** [typescript-language-server/typescript-language-server#909] [Q&A]
Hi, 
As we know LSP is implement by Client and Server, the `typescript-language-server` is the Server.

and generally speaking the Client is implement by IDE(such as VS Code).

and when we use  `typescript-language-server`, we just call the API of `typescript-language-server`, that's it, right? don't need a Client.

**Comment by rchl** [comment]
It's intended to be used by communicating with it using LSP protocol. Not through programmatic API.

**Reply by aabounegm** [reply]
@rchl Is there a way (or are you open to the possibility) to expose and use a programmatic API? I am developing an LSP server for a language that can have embedded JavaScript in it, and I wanted to use the [Language Services](https://code.visualstudio.com/api/language-extensions/embedded-languages#language-services) approach to redirect requests in JavaScript portions to the JS language service.
If this cannot be done with this package, can you point me to a library that supports that?

**Reply by rchl** [reply]
I think that https://github.com/volarjs/volar.js (https://volarjs.dev/) is basically designed for your use case.

https://github.com/vuejs/language-tools and some other frameworks use Volar so those can possibly be used for inspiration.

**Reply by aabounegm** [reply]
Thanks for the reply. I checked out Volar recently but the documentation was not comprehensive enough and I didn't find many example projects (only Vue and Astro), so I decided not to use it, but maybe it's worth exploring more closely. I guess I should give it a shot
